asm_to_pil/src/vm_to_constrained.rs

```diff
@@ -0,0 +1,995 @@
+    ) {
+        let mut conditioned_updates = vec![];
+        let mut default_update = None;
+        match ty {
+            RegisterTy::Pc => {
+                assert_eq!(self.pc_name, None);
+                self.pc_name = Some(name.to_string());
+                self.line_lookup
+                    .push((name.to_string(), "p_line".to_string()));
+                default_update = Some(build_add(direct_reference(&name), build_number(1u64)));
+            }
+            RegisterTy::Assignment => {
+                // no default update as this is transient
+            }
+            RegisterTy::ReadOnly => {
+                // default update to be kept constant
+                default_update = Some(direct_reference(&name))
+            }
+            RegisterTy::Write => {
+                let assignment_regs = self
+                    .assignment_register_names()
+                    .cloned()
+                    .collect::<Vec<_>>();
+                // TODO do this at the same place where we set up the read flags.
+                for reg in assignment_regs {
+                    let write_flag = format!("reg_write_{reg}_{name}");
+                    self.create_witness_fixed_pair(start, &write_flag);
+                    conditioned_updates
+                        .push((direct_reference(&write_flag), direct_reference(&reg)));
+                }
+                default_update = Some(direct_reference(&name));
+            }
+        };
+        self.registers.insert(
+            name.to_string(),
+            Register {
+                conditioned_updates,
+                default_update,
+                ty,
+            },
+        );
+        self.pil.push(witness_column(start, name, None));
+    }
+
+    fn handle_instruction_def(
+        &mut self,
+        s: InstructionDefinitionStatement<T>,
+    ) -> Option<InstructionDefinitionStatement<T>> {
+        let instruction_name = s.name.clone();
+        let instruction_flag = format!("instr_{instruction_name}");
+        self.create_witness_fixed_pair(s.start, &instruction_flag);
+        // it's part of the lookup!
+        //self.pil.push(constrain_zero_one(&col_name));
+
+        let inputs: Vec<_> = s
+            .instruction
+            .params
+            .clone()
+            .inputs
+            .params
+            .into_iter()
+            .map(|param| match param.ty {
+                Some(ty) if ty == "label" => Input::Literal(param.name, LiteralKind::Label),
+                Some(ty) if ty == "signed" => {
+                    Input::Literal(param.name, LiteralKind::SignedConstant)
+                }
+                Some(ty) if ty == "unsigned" => {
+                    Input::Literal(param.name, LiteralKind::UnsignedConstant)
+                }
+                None => Input::Register(param.name),
+                Some(ty) => panic!("param type must be nothing or label, found `{ty}`"),
```

Comment:
what about signed and unsigned above?

Reply:
Removed as this should be done in the type checker